internal/app/machined/pkg/runtime/v1alpha1/v1alpha1_controller.go

```diff
@@ -92,59 +118,53 @@ func (c *Controller) Run(ctx context.Context, seq runtime.Sequence, data interfa
 		}
 	}
 
-	// Allow only one sequence to run at a time with the exception of bootstrap
-	// and reset sequences.
-	switch seq { //nolint:exhaustive
-	case runtime.SequenceReset:
-		// Do not attempt to lock.
-	default:
-		if opts.Force {
-			break
+	canTakeOver := true
+
+	c.ctxMutex.Lock()
+
+	if c.runningSequence != -1 {
+		canTakeOver = false
+
+		config, ok := sequenceTakeOver[c.runningSequence]
+		if ok {
+			_, canTakeOver = config[seq]
 		}
 
-		if opts.Takeover {
-			c.ctxMutex.Lock()
-			if c.cancelCtx != nil {
-				c.cancelCtx()
-			}
+		canTakeOver = canTakeOver || opts.Takeover
+	}
 
-			c.ctxMutex.Unlock()
+	if canTakeOver && c.cancelCtx != nil {
+		c.cancelCtx()
+	}
 
-			err := retry.Constant(time.Minute*1, retry.WithUnits(time.Millisecond*100)).RetryWithContext(ctx, func(ctx context.Context) error {
-				if c.TryLock() {
-					return retry.ExpectedError(fmt.Errorf("failed to acquire lock"))
-				}
+	c.ctxMutex.Unlock()
 
-				return nil
-			})
-			if err != nil {
-				return err
-			}
-		} else if c.TryLock() {
-			c.Runtime().Events().Publish(&machine.SequenceEvent{
-				Sequence: seq.String(),
-				Action:   machine.SequenceEvent_NOOP,
-				Error: &common.Error{
-					Code:    common.Code_LOCKED,
-					Message: fmt.Sprintf("sequence not started: %s", runtime.ErrLocked.Error()),
-				},
-			})
-
-			return runtime.ErrLocked
+	if !canTakeOver {
+		return runtime.ErrLocked
+	}
+
+	err := retry.Constant(time.Minute*1, retry.WithUnits(time.Millisecond*100)).RetryWithContext(ctx, func(ctx context.Context) error {
+		if c.TryLock() {
+			return retry.ExpectedError(fmt.Errorf("failed to acquire lock"))
 		}
 
-		defer c.Unlock()
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	defer c.Unlock()
```

Comment:
there is a gap here with the flow and decision making:

1. first we check whether we can take over the running sequence (while holding `c.ctxMutex`)
2. after the decision is made we attempt to acquire another lock with `c.TryLock`, but beween 1 and 2 there might be any change to the running sequence, or two concurrent `Run()` might both make a decision to take over (while they can't takeover each other).
3. same with canceling the running sequence - it looks like several concurrent `Run()` calls can do that.


It feels to me we should have a much broader and proper lock around the decision to run the sequence, cancel the running, etc. In other words, if this `Run()` is going to take over a running sequence, it should do it so that no other goroutine can do the same. 

The smallest possible change is to re-check the conditions after making a final decisions, but this is not clear for me.